File: 106.construct-binary-tree-from-inorder-and-postorder-traversal.js
/*
 * @lc app=leetcode id=106 lang=javascript
 *
 * [106] Construct Binary Tree from Inorder and Postorder Traversal
 */

function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val;
  this.left = left === undefined ? null : left;
  this.right = right === undefined ? null : right;
}

// @lc code=start
/**
 * @param {number[]} inorder
 * @param {number[]} postorder
 * @return {TreeNode}
 */
var buildTree = function (inorder, postorder) {
  const helper = (p, iLeft, iRight) => {
    if (p < 0 || iLeft > iRight) {
      return null;
    }
    const i = lookup[postorder[p]];
    const root = new TreeNode(inorder[i]);
    root.left = helper(p - iRight + i - 1, iLeft, i - 1);
    root.right = helper(p - 1, i + 1, iRight);
    return root;
  };
  const lookup = {};
  for (let i = 0; i < inorder.length; ++i) {
    lookup[inorder[i]] = i;
  }
  return helper(postorder.length - 1, 0, inorder.length - 1);
};
// @lc code=end

File: 106.construct-binary-tree-from-inorder-and-postorder-traversal.py
#
# @lc app=leetcode id=106 lang=python3
#
# [106] Construct Binary Tree from Inorder and Postorder Traversal
#

# Time:  O(n)
# Space: O(n)

from typing import List


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


# @lc code=start
class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> TreeNode:
        def helper(start, end):
            if start > end:
                return None
            node_val = next(postorder_iter)
            node = TreeNode(node_val)
            idx = inorder_lookup[node_val]
            node.right = helper(idx + 1, end)
            node.left = helper(start, idx - 1)
            return node

        postorder_iter = iter(postorder[::-1])
        inorder_lookup = {n: i for i, n in enumerate(inorder)}
        return helper(0, len(inorder) - 1)


# @lc code=end
